src/crypto/cng.c

```diff
@@ -418,10 +419,13 @@ _cng_ctr_crypto_generate (cng_ctr_encrypt_state *state,
 static void
 _cng_ctr_crypto_advance (cng_ctr_encrypt_state *state)
 {
-   BSON_ASSERT (state);
+   BSON_ASSERT_PARAM (state);
+
+   /* Assert rather than return false/NULL since this function's type is void */
+   BSON_ASSERT (state->input_block_len <= INT_MAX);
 
    uint32_t carry = 1;
-   for (int i = state->input_block_len - 1; i >= 0 && carry != 0; --i) {
+   for (int i = (int) state->input_block_len - 1; i >= 0 && carry != 0; --i) {
```

Comment:
Because `state->input_block_len` is a uint32_t, there's a possibility for signed overflow when converting it to an int. Instead of casting to an int, perhaps `i` could be a `uint32_t` to match the type of `input_block_len` in order to avoid a possible integer overflow scenario?

i.e.
```C
   for (uint32_t i = state->input_block_len - 1; i >= 0 && carry != 0; --i) {
```

Reply:
Ok, I think I may be wrong on that last point, I think this would be an instance of implementation defined behavior of unsigned to signed conversion (therefor not undefined behavior). Given all the platforms we support are two's complement, then this shouldn't cause a problem!

Reference:
https://stackoverflow.com/a/67755393

Reply:
Excellent! Thanks for pointing that out. I hadn't noticed that on first review.